### PR TITLE
Update: no-inner-declarations false negative in non-block (fixes #12222)

### LINKED ISSUE #12222: no-inner-declarations inconsistency
**Tell us about your environment**

* **ESLint Version:**  6.3.0
* **Node Version:** 10.16.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  parserOptions: {
    ecmaVersion: 2015,
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

[Demo link](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8taW5uZXItZGVjbGFyYXRpb25zOiBbXCJlcnJvclwiLCBcImJvdGhcIl0qL1xuXG5pZiAoZm9vKVxuICB2YXIgYTtcblxuaWYgKGZvbylcbiAgZnVuY3Rpb24gZigpe31cblxuZnVuY3Rpb24gYmFyKCkge1xuICBpZiAoZm9vKVxuICAgIHZhciBhO1xuXG4gIGlmIChmb28pXG4gICAgZnVuY3Rpb24gZigpe31cbn1cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS1hdG9taWMtdXBkYXRlcyI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)

```js
/*eslint no-inner-declarations: ["error", "both"]*/

if (foo)
  var a;

if (foo)
  function f(){}

function bar() {
  if (foo)
    var a;

  if (foo)
    function f(){}
}
```

```bash
eslint index.js
```

**What did you expect to happen?**

To be consistent, either 4 errors or no errors?

> This rule requires that function declarations and, optionally, variable declarations be in the **root** of a program or the **body** of a function.

Are these declarations in the root/body? Also, the documentation is later more specific:

> "functions" (default) disallows function declarations in nested **blocks**

> "both" disallows function and var declarations in nested **blocks**

**What actually happened? Please include the actual, raw output from ESLint.**

2 errors:

```
11:5  error  Move variable declaration to function body root  no-inner-declarations
14:5  error  Move function declaration to function body root  no-inner-declarations
```

```js
/*eslint no-inner-declarations: ["error", "both"]*/

if (foo)
  var a; // no error

if (foo)
  function f(){} // no error

function bar() {
  if (foo)
    var a; // error

  if (foo)
    function f(){} // error
}
```

**Are you willing to submit a pull request to fix this bug?**

Yes, if this is confirmed to be a bug.

Perhaps the rule should allow only these places for `function` and `var` declaration nodes:

* `Program > node`
* `ExportNamedDeclaration > node`
* `ExportDefaultDeclaration > node` (function only)
* `Function > BlockStatement > node`

meaning that there should be 4 errors in the above example, but I'm not sure is this the intention of this rule.

### PR DESCRIPTION
**Pull request by anikethsaha**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [X] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

fixes https://github.com/eslint/eslint/issues/12222

#### What changes did you make? (Give an overview)
Currently whenever the distance is more than 1, it was not showing error cause it was considering ancestor as any (`Program` included). 
it will be valid only if ancestor is not `Program` for distance equal to 2

example

```js
if(foo){
  var a;
}
```
this will result in distance 1 for the ancestor `Program`

```js
if(foo)
  var a;
```
This will be distance 2 for the ancestor `Program`. 
here both calculations for distance and ancestor are correct but not valid. (the `master` is considering it as valid)

So in this PR, whenever distance is 2, it should not check if from `Program`.

#### Is there anything you'd like reviewers to focus on?
#12222

### CODE REVIEW

**Comment by mdjermanovic** [comment]
These should be also allowed:

```js
/*eslint no-inner-declarations: ["error", "both"]*/

export var foo;

export function bar() {}

export default function baz() {}
```

**Comment by anikethsaha** [comment]
cc @mdjermanovic 

I changed the validation condition, now its allowing those named and default exports

**Comment by mdjermanovic** [comment]
Thoughts about simplifying the code?

A declaration is okay if its parent is either of:

* Program
* ExportNamedDeclaration
* ExportDefaultDeclaration
* BlockStatement which parent is either of FunctionDeclaration, FunctionExpression or ArrowFunctionExpression.

so maybe we could just check these simple conditions, instead of calculating the distance and then analyzing the result.

**Comment by anikethsaha** [comment]
@mdjermanovic I simplified the code as per your suggestion and it is working as intended !

that while loop was needed as we need to find the correct immediate ancestor where the declaration should be moved for the error message. 

let me know if it needs some more changes 👍

**Comment by mdjermanovic** [comment]
> that while loop was needed as we need to find the correct immediate ancestor where the declaration should be moved for the error message.

You're right! I missed the fact that the ancestor is used for the error message as well.

File: lib/rules/no-inner-declarations.js [hunk 1]
```diff
@@ -34,53 +37,48 @@ module.exports = {
...
+            let ancestor = ancestors.pop();
 
-            while (ancestor && ["Program", "FunctionDeclaration",
-                "FunctionExpression", "ArrowFunctionExpression"
-            ].indexOf(ancestor.type) < 0) {
-                generation += 1;
-                ancestor = ancestors.pop();
+            if (
+                ancestor.type === "BlockStatement" &&
+            !!~validBlockStatementParent.indexOf(ancestor.parent.type)
```
Comment (mdjermanovic) [inline lib/rules/no-inner-declarations.js]: Can we use `validBlockStatementParent.includes` here, I think it's more readable than `!!~validBlockStatementParent.indexOf`.

Alternatively, `validBlockStatementParent` might be a `Set` instead of an array.

(the same applies to the other condition below)

File: lib/rules/no-inner-declarations.js [hunk 2]
```diff
@@ -34,53 +37,48 @@ module.exports = {
...
-                    }
-                });
+            while (ancestor && !~parentsListToCheck.indexOf(ancestor.type)) {
+                ancestor = ancestors.pop();
+                if (ancestor.type === "BlockStatement") {
+                    ancestor = ancestor.parent;
+                    parentsListToCheck = validBlockStatementParent;
+                } else {
+                    parentsListToCheck = validParent;
+                }
```
Comment (mdjermanovic) [inline lib/rules/no-inner-declarations.js]: Can we simplify this, I think the goal is just to find the lowest ancestor that is either `Program`, `FunctionDeclaration`, `FunctionExpression` or `ArrowFunctionExpression`. This seems unrelated to `validParent` and `validBlockStatementParent`.

**Comment by anikethsaha** [comment]
@mdjermanovic 
should I revert the latest commit then? the latest commit gives the ancestor as expected as the old one  though !

File: lib/rules/no-inner-declarations.js (hunk 2 above)
Comment (mdjermanovic) [reply lib/rules/no-inner-declarations.js]: We could maybe use `astUtils.getUpperFunction`.

File: lib/rules/no-inner-declarations.js (hunk 1 above)
Comment (anikethsaha) [reply lib/rules/no-inner-declarations.js]: > Alternatively, `validBlockStatementParent` might be a `Set` instead of an array.

This seems much better 👍 I will change it soon

File: lib/rules/no-inner-declarations.js (hunk 2 above)
Comment (anikethsaha) [reply lib/rules/no-inner-declarations.js]: > We could maybe use `astUtils.getUpperFunction`.

didn't know about this, I will take a look 👍

Comment (anikethsaha) [reply lib/rules/no-inner-declarations.js]: done

File: lib/rules/no-inner-declarations.js (hunk 1 above)
Comment (anikethsaha) [reply lib/rules/no-inner-declarations.js]: done

**Comment by anikethsaha** [comment]
@mdjermanovic I updated the docs with some examples

File: docs/rules/no-inner-declarations.md [hunk 3]
```diff
@@ -137,6 +137,10 @@ if (test) {
 function doAnotherThing() {
     var baz = 81;
 }
+
+if (foo) var a;
+
+if (foo) function f(){}
```
Comment (mdjermanovic) [inline docs/rules/no-inner-declarations.md]: These should be in "incorrect examples".

Maybe `if (foo) function f(){}` in incorrect for "functions", `if (foo) var a;` in incorrect for "both".

File: tests/lib/rules/no-inner-declarations.js [hunk 4]
```diff
@@ -45,74 +45,235 @@ ruleTester.run("no-inner-declarations", rule, {
...
+        {
+            code: "exports.foo = function(){}",
+            options: ["both"],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "module.exports = function foo(){}",
+            options: ["both"],
+            parserOptions: { ecmaVersion: 6 }
         }
```
Comment (mdjermanovic) [inline tests/lib/rules/no-inner-declarations.js]: `ecmaVersion: 6` is not necessary for these two tests.

File: docs/rules/no-inner-declarations.md (hunk 3 above)
Comment (anikethsaha) [reply docs/rules/no-inner-declarations.md]: oh shoot, I added them in wrong section. I will change it

File: tests/lib/rules/no-inner-declarations.js (hunk 4 above)
Comment (anikethsaha) [reply tests/lib/rules/no-inner-declarations.js]: my bad ! 👍

**Comment by anikethsaha** [comment]
@mdjermanovic done 👍

**Comment by kaicataldo** [comment]
Removing the "needs bikeshedding" label as per [this comment](https://github.com/eslint/eslint/issues/12222#issuecomment-606225113).
